File: 中等/参考了/56合并区间.py
"""

给出一个区间的集合，请合并所有重叠的区间。

示例 1:
输入: intervals = [[1,3],[2,6],[8,10],[15,18]]
输出: [[1,6],[8,10],[15,18]]
解释: 区间 [1,3] 和 [2,6] 重叠, 将它们合并为 [1,6].

示例2:
输入: intervals = [[1,4],[4,5]]
输出: [[1,5]]
解释: 区间 [1,4] 和 [4,5] 可被视为重叠区间。

提示：
intervals[i][0] <= intervals[i][1]

"""


# 基本思路
# 我们用数组 merged 存储最终的答案。
# 首先，我们将列表中的区间按照左端点升序排序。
# 然后我们将第一个区间加入 merged 数组中，并按顺序依次考虑之后的每个区间：
# 如果当前区间的左端点在数组 merged 中最后一个区间的右端点之后，那么它们不会重合，
# 我们可以直接将这个区间加入数组 merged 的末尾；
# 否则，它们重合，我们需要用当前区间的右端点更新数组 merged 中最后一个区间的右端点，将其置为二者的较大值。
def merge(intervals):
    intervals.sort()
    merged = []
    merged.append(intervals[0])
    for i in range(1, len(intervals)):
        if merged[-1][1] < intervals[i][0]:
            merged.append(intervals[i])
        else:
            if merged[-1][1] < intervals[i][1]:
                merged[-1][1] = intervals[i][1]
    return merged


print(merge([[1, 2], [3, 5], [6, 7], [8, 10], [12, 16], [4, 8]]))

File: 剑指Offer Ⅱ/56. 合并区间.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  const res = intervals.sort((a, b) => {
    return a[0] - b[0]
  })
  let i = 1
  while (i < res.length) {
    if (res[i - 1][1] < res[i][0]) {
      i += 1
      continue
    } else if (res[i - 1][1] > res[i][0]) {
      if (res[i - 1][1] >= res[i][1]) {
        res.splice(i, 1)
      } else {
        const temp = [res[i - 1][0], res[i][1]]
        res.splice(i - 1, 2)
        res.splice(i - 1, 0, temp)
      }
    } else {
      const temp = [res[i - 1][0], res[i][1]]
      res.splice(i - 1, 2)
      res.splice(i - 1, 0, temp)
    }
  }
  return res
};


console.log(merge([[1, 3], [2, 6], [8, 10], [15, 18]]));
console.log(merge([[1, 4], [4, 5]]));


// 二维数组排序
// intervals.sort((a, b) => a[0] - b[0]);//按照起始位置排序
// const res = [intervals[0]]

// for (let i = 1; i < intervals.length; i++) {
//   if (intervals[i][0] > res[res.length - 1][1]) {
//     res.push(intervals[i])
//     continue
//   }
//   else if (intervals[i][0] <= res[res.length - 1][1] && intervals[i][0] > res[res.length - 1][0]) {
//     if (intervals[i][1] > res[res.length - 1][1]) {
//       const temp = [res[res.length - 1][0], intervals[i][1]]
//       res.pop()
//       res.push(temp)
//     } else {
//       const temp = [res[res.length - 1][0], res[res.length - 1][1]]
//       res.pop()
//       res.push(temp)
//     }
//   } else if (intervals[i][0] <= res[res.length - 1][0]) {
//     if (res[res.length - 1][1] <= intervals[i][1]) {
//       const temp = [intervals[i][0], intervals[i][1]]
//       res.pop()
//       res.push(temp)
//     } else {
//       const temp = [intervals[i][0], res[res.length - 1][1]]
//       res.pop()
//       res.push(temp)
//     }
//   }
// }
// return res